src/commands/signzone.rs

```diff
@@ -0,0 +1,285 @@
+    origin: Option<Name<Bytes>>,
+
+    /// The zonefile to sign
+    #[arg(value_name = "zonefile")]
+    zonefile_path: PathBuf,
+
+    /// The keys to sign the zone with
+    ///
+    /// keys must be specified by their base name (usually
+    /// K<name>+<alg>+<id>), i.e. WITHOUT the .private extension.
+    #[arg(value_name = "key")]
+    key_paths: Vec<PathBuf>,
+}
+
+impl SignZone {
+    pub fn execute(self) -> Result<(), Error> {
+        let mut records = self.load_zone()?;
+
+        //---
+
+        // Import the specified key.
+        let mut keys = vec![];
+        for key_path in self.key_paths {
+            let old_ext = key_path.extension().unwrap().to_str().unwrap();
+            let new_ext = format!("{}.private", old_ext);
+            let private_key_path = key_path.with_extension(new_ext).display().to_string();
+            let private_data = std::fs::read_to_string(&private_key_path).map_err(|err| {
+                Error::from(format!(
+                    "Unable to load private key from file '{}': {}",
+                    private_key_path, err
+                ))
+            })?;
+
+            let old_ext = key_path.extension().unwrap().to_str().unwrap();
+            let new_ext = format!("{}.key", old_ext);
+            let public_key_path = key_path.with_extension(new_ext).display().to_string();
+            let public_data = std::fs::read_to_string(&public_key_path).map_err(|err| {
+                Error::from(format!(
+                    "Unable to load public key from file '{}': {}",
+                    private_key_path, err
+                ))
+            })?;
+
+            let secret_key =
+                domain::sign::SecretKeyBytes::parse_from_bind(&private_data).map_err(|err| {
+                    Error::from(format!(
+                        "Unable to parse BIND formatted private key file '{}': {}",
+                        private_key_path, err
+                    ))
+                })?;
+
+            let public_key_info: domain::validate::Key<Bytes> =
+                domain::validate::Key::parse_from_bind(&public_data).map_err(|err| {
+                    Error::from(format!(
+                        "Unable to parse BIND formatted public key file '{}': {}",
+                        private_key_path, err
+                    ))
+                })?;
+
+            let key_pair = domain::sign::common::KeyPair::from_bytes(
+                &secret_key,
+                public_key_info.raw_public_key(),
+            )
+            .map_err(|err| {
+                Error::from(format!(
+                    "Unable to import private key from file '{}': {}",
+                    private_key_path, err
+                ))
+            })?;
+
+            let signing_key = domain::sign::SigningKey::new(
+                public_key_info.owner().to_owned(),
+                public_key_info.flags(),
+                key_pair,
+            );
+
+            keys.push(signing_key);
```

Comment:
Perhaps we want to move the above loading logic into a separate function here, and then just call `key_paths.into_iter().map(...).try_collect()?`.

Reply:
Until now I hadn't done that because the key management code was still evolving and I didn't want to clean up too much until that stabilized. But that's merged now so yes I can extract this. The same btw goes for the hashing/signing code, I'm happy to leave that verbose/messy while the underlying library code is still changing.

Reply:
`try_collect()` appears to be unstable.

Reply:
I factored out the logic to a helper function but didn't use a functional iterator style collect-to-vec due to not being able to use the ? operator in a closure. Instead I used a simple for loop.